docs/en/week12/12-3.md

```diff
@@ -0,0 +1,239 @@
+Figure 14: Cost to every step
+</center>
+
+$(1)$ Cost Example: Average Distance
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure15.png" style="background-color:#DCDCDC;" /><br>
+Figure 15: Cost Example: Average Distance
+</center>
+
+$(2)$ Cost Example: Softmin
+<center>
+<img src="{{site.baseurl}}/images/week12/12-3/figure16.png" style="background-color:#DCDCDC;" /><br>
+Figure 16: Cost Example: Softmin
+</center>
+
+Different forms of cost functions can be explored through experimentation. 
+## Optimization_Path_Planner-Notebook
+In this notebook, we use tri-cycle as an example as well. 
+
+### Define kinematic model of a tricycle $\dot{x}=f(x,u)$. 
+* $x$ represents state: ($x$, $y$, $θ$, $s$)
+* $u$ represents control: ($ϕ$, $a$)
+* We feed $x[t-1]$ and $u[t]$ to obtain the next state $x[t]$
+
+
+```
+def f(x, u, t=None):
+    L = 1  # m
+    x, y, θ, s = x
+    
+    ϕ, a = u
+    f = torch.zeros(4)
+    f[0] = s * torch.cos(θ)
+    f[1] = s * torch.sin(θ)
+    f[2] = s / L * torch.tan(ϕ)
+    f[3] = a
+    return f
+```
+### Define several cost functions
+As mentioned above, cost functions can take various forms. In this notebook, we list 5 kinds as follows: 
+* **vanilla_cost**: Focuses on the final position
+* **cost_with_target_s**: Focuses on the final position and final zero speed. 
+* **cost_sum_distances**: Focuses on the position of every step, and minimizes the mean of the distances. 
+* **cost_sum_square_distances**: Focuses on the position of every step, and minimizes the mean of squared distances.
+* **cost_logsumexp**: The distance of the closest position should be minimized. 
+
+
+```
+def vanilla_cost(state, target):
+    x_x, x_y = target
+    return (state[-1][0] - x_x).pow(2) + (state[-1][1] - x_y).pow(2)
+
+def cost_with_target_s(state, target):
+    x_x, x_y = target
+    return (state[-1][0] - x_x).pow(2) + (state[-1][1] - x_y).pow(2) + (state[-1][-1]).pow(2)
+
+def cost_sum_distances(state, target):
+    x_x, x_y = target
+    dists = ((state[:, 0] - x_x).pow(2) + (state[:, 1] - x_y).pow(2)).pow(0.5)
+    return dists.mean()
+
+def cost_sum_square_distances(state, target):
+    x_x, x_y = target
+    dists = ((state[:, 0] - x_x).pow(2) + (state[:, 1] - x_y).pow(2))
+    return dists.mean()
+
+def cost_logsumexp(state, target):
+    x_x, x_y = target
+    dists = ((state[:, 0] - x_x).pow(2) + (state[:, 1] - x_y).pow(2))#.pow(0.5)
+    return -1 * torch.logsumexp(-1 * dists, dim=0)
+```
+### Define path planning with cost
+* The optimizer is set to be SGD.
+* Time interval T is set to be 1s.
+* We need to compute every state from the initial state by the following code: 
+```
+x = [torch.tensor((0, 0, 0, s),dtype=torch.float32)]
+        for t in range(1, T+1):
+            x.append(x[-1] + f(x[-1], u[t-1]) * dt)
+        x_t = torch.stack(x)
+```
+* Then compute the cost:
+  `cost = cost_f(x_t, (x_x, x_y))
+        costs.append(cost.item())`
```

Comment:
This looks like this. I'm not sure what you were aiming for.
![image](https://user-images.githubusercontent.com/2119355/117222726-2333b880-adda-11eb-8d5a-6a1ea99314f1.png)

Reply:
After we choose a specific cost (vanilla, average distance...), we can compute the cost, which is the distance of our target (x_x, x_y) and our positions x_t. The term "costs.append..." can be deleted.

Reply:
I meant I don't know how you wanted it to appear.
I see a in-line code split in two lines.
I didn't know whether you wanted to use a code-block or something else.

Reply:
I wanted to use a code-block here. I am not sure how it appears like this. I'll figure it out.

Reply:
Just use three backticks rather than a single one.